ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs

```diff
@@ -17,6 +17,121 @@
 {-# LANGUAGE TupleSections              #-}
 {-# LANGUAGE TypeApplications           #-}
 
+{-|
+  = Snapshot managing and contents
+
+  Snapshotting a ledger state means saving a copy of the in-memory part of the
+  ledger state serialized as a file in the disk, as well as flushing differences
+  between the last snapshotted ledger state and the one that we are snapshotting
+  now onto the BackingStore. While the coupled implementation is in place, the
+  serialized file might also contain a serialized version of the in-memory UTxO
+  map carried by the legacy ledger states, as explained below.
+
+  == Startup
+
+  On startup, the node will:
+
+  1. Find the latest snapshot which will be a directory identified by the slot
+     number of the snapshot:
+
+        > <cardano-node data dir>
+        > ├── volatile
+        > ├── immutable
+        > └── ledger
+        >     ├── <slotNumber1>
+        >     │   ├── tables
+        >     │   └── state
+        >     ├── <slotNumber2>
+        >     │   ├── tables
+        >     │   └── state
+        >     └── <slotNumber3>
+        >         ├── tables
+        >         └── state
+
+  2. Depending on the snapshots found, there are two paths:
+
+       - If there is no snapshot to load, create a new @BackingStore@ with the
+         contents of the Genesis UTxO and finish.
+
+       - If there is a snapshot found, then deserialize the @state@ file, which
+         contains either:
+
+            - If the node was running __with the legacy database enabled__, it
+              contains a full LedgerState snapshot in the same style as before
+              the UTxO HD implementation (i.e. in-memory part of the LedgerState
+              (equal among implementations) + UTxO map).
+
+            - Otherwise, it contains a serialization of the in-memory part of the
+              LedgerState, without a UTxO map.
+
+        In both cases, the snapshot will be deserialized as @ExtLedgerState blk
+        EmptyMK@. In case we found an snapshot, we will overwrite the
+        @BackingStore@ tables with the @tables@ file from said snapshot.
+
+  3. Depending on whether the node is requested to run both databases and whether
+     or not the snapshot has a UTxO inside, the node will:
+
+       - __Run both__ + snapshot has UTxO inside: Extract the tables from the
+         deserialized snapshot to create a @ExtLedgerState blk ValuesMK@ which
+         can be 1. used in the legacy database and 2. after stripping out the
+         @ValuesMK@, used in the modern database.
+
+       - __Run both__ + snapshot doesn't have UTxO inside: an error will be
+         thrown as we cannot/don't think we should reconstruct an in-memory UTxO
+         directly from the @BackingStore@.
+
+       - Otherwise: Strip out the values inside the @ExtLedgerState blk EmptyMK@
+         to make sure that we don't have an in-memory UTxO map and use it as the
+         anchor for the modern database.
```

Comment:
How can we distinguish the case in which we have an empty UTxO map just because there are no transactions (admittedly a pathological case) and the case in which the map is empty because we didn't store the legacy ledger state?

Reply:
The UTxO map would be the one from the legacy ledger and the only possible case in which it would be empty is if every account had a balance of `0` which hopefully does never happen.